server/protocols/webadmin/webadmin-mailbox-deleted-message-vault/src/main/java/org/apache/james/webadmin/vault/routes/ExportService.java

```diff
@@ -0,0 +1,128 @@
+import com.github.fge.lambdas.Throwing;
+import com.github.fge.lambdas.functions.ThrowingFunction;
+import com.github.steveash.guavate.Guavate;
+import com.google.common.annotations.VisibleForTesting;
+
+import reactor.core.publisher.Flux;
+import reactor.core.publisher.Mono;
+import reactor.core.scheduler.Schedulers;
+
+class ExportService {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(ExportService.class);
+
+    private final BlobExportMechanism blobExport;
+    private final BlobStore blobStore;
+    private final DeletedMessageZipper zipper;
+    private final DeletedMessageVault vault;
+
+    @Inject
+    @VisibleForTesting
+    ExportService(BlobExportMechanism blobExport, BlobStore blobStore, DeletedMessageZipper zipper, DeletedMessageVault vault) {
+        this.blobExport = blobExport;
+        this.blobStore = blobStore;
+        this.zipper = zipper;
+        this.vault = vault;
+    }
+
+    Mono<Void> export(User user, Query exportQuery, MailAddress exportToAddress,
+                      Runnable messageToExportCallback) {
+
+        return matchingMessages(user, exportQuery)
+            .doOnNext(any -> messageToExportCallback.run())
+            .collect(Guavate.toImmutableList())
+            .map(Collection::stream)
+            .map(sneakyThrow(messages -> zipData(user, messages)))
+            .flatMap(sneakyThrow(zippedStream -> blobStore.save(zippedStream, zippedStream.available())))
+            .flatMap(blobId -> exportTo(user, exportToAddress, blobId))
+            .then();
+    }
+
+    private Flux<DeletedMessage> matchingMessages(User user, Query exportQuery) {
+        return Flux.from(vault.search(user, exportQuery))
+            .publishOn(Schedulers.elastic());
+    }
+
+    private PipedInputStream zipData(User user, Stream<DeletedMessage> messages) throws IOException {
+        PipedOutputStream outputStream = new PipedOutputStream();
+        PipedInputStream inputStream = new PipedInputStream();
+        inputStream.connect(outputStream);
+
+        asyncZipData(user, messages, outputStream).subscribe();
+
+        return inputStream;
+    }
+
+    private Mono<Void> asyncZipData(User user, Stream<DeletedMessage> messages, PipedOutputStream outputStream) {
+        return Mono.fromRunnable(Throwing.runnable(() -> zipper.zip(message -> loadMessageContent(user, message), messages, outputStream)).sneakyThrow())
+            .doOnSuccessOrError(Throwing.biConsumer((result, throwable) -> {
+                if (throwable != null) {
+                    LOGGER.error("Error happens when zipping deleted messages", throwable);
+                }
+                outputStream.flush();
+                outputStream.close();
+            }))
+            .subscribeOn(Schedulers.elastic())
+            .then();
+    }
+
+    private InputStream loadMessageContent(User user, DeletedMessage message) {
+        return Mono.from(vault.loadMimeMessage(user, message.getMessageId()))
+            .block();
```

Comment:
IDK how I should handle the emptyness, when MessageLoader return an InputStream instead of Optional or Mono...

Reply:
and one more concern need to be validated here